cpp/include/cudf/binning/bin.hpp

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/types.hpp>
+
+#include <rmm/mr/device/device_memory_resource.hpp>
+#include <rmm/mr/device/per_device_resource.hpp>
+
+namespace cudf {
+
+/**
+ * @addtogroup binning
+ * @{
+ * @file
+ * @brief APIs for binning values.
+ */
+
+/**
+ * @brief Enum used to define whether or not bins include their boundary points.
+ */
+enum class inclusive { YES, NO };
+
+/**
+ * @brief Labels elements based on membership in the specified bins.
+ *
+ * A bin `i` is defined by `left_edges[i], right_edges[i]`. Whether the edges are inclusive or
+ * not is determined by `left_inclusive` and `right_inclusive`, respectively. If either
+ * `left_edges[i]` or `right_edges[i]` is NULL, no value in `input` can be placed in bin `i`.
+ *
+ * A value `input[j]` belongs to bin `i` if `value[j]` is contained in the range `left_edges[i],
+ * right_edges[i]` (with the specified inclusiveness) and `label[j] == i`. If  `input[j]` does not
+ * belong to any bin, then `label[j]` is NULL.
+ *
+ * Notes:
+ *   - If an empty set of edges is provided, all elements in `input` are labeled NULL.
+ *   - NULL elements in `input` belong to no bin and their corresponding label is NULL.
+ *   - NaN elements in `input` belong to no bin and their corresponding label is NULL.
+ *   - Bins must be provided in monotonically increasing order, otherwise behavior is undefined.
+ *   - If two or more bins overlap, behavior is undefined.
+ *   - If nulls are interspersed throughout the edges rather than being at the beginning or end
+ *     (as set by @p edge_null_precedence), behavior is undefined.
```

Comment:
This strikes me as odd. What's the usecase to allow this? If the user knows all of the nulls are contiguous they could do a slice to just give us the non-null elements.

Reply:
The original API spec we were working with didn't specify anything regarding the behavior of the function with respect to nulls in the edges. It was pointed out to me that very few places in cudf _can't_ handle nulls, but also that for sorted inputs the nulls will be contiguously placed at either the beginning or the end of the column, so this seemed like an easy use-case to support. I'd also be fine just throwing a logic error if you think that's a more sensible choice here.

Reply:
Fair enough. That simplifies this code a bit. Should I bother to throw a logic error if the edges contain nulls, or does this fall under "libcudf doesn't make decision" and we just document null edges as leading to UB?

Reply:
We always check for nulls in a situation where nulls are not allowed. I can see how one could make the argument that this is like making a decision, but in this case I think it's a good decision.

Reply:
That sounds like a reasonable choice to me.